src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs

```diff
@@ -132,46 +132,89 @@ internal static void SetDate(this IHeaderDictionary headers, string name, DateTi
+
+    private static IList<StringWithQualityHeaderValue> ParseStringWithQualityHeaderValue(IList<string> value) =>
+        StringWithQualityHeaderValue.TryParseList(value, out var result) ? result : Array.Empty<StringWithQualityHeaderValue>();
+
+    private static IList<CookieHeaderValue> ParseCookieHeaderValue(IList<string> value) =>
+        CookieHeaderValue.TryParseList(value, out var result) ? result : Array.Empty<CookieHeaderValue>();
+
+    private static IList<EntityTagHeaderValue> ParseEntityTagHeaderValue(IList<string> value) =>
+        EntityTagHeaderValue.TryParseList(value, out var result) ? result : Array.Empty<EntityTagHeaderValue>();
+
+    private static IList<SetCookieHeaderValue> ParseSetCookieHeaderValue(IList<string> value) =>
+        SetCookieHeaderValue.TryParseList(value, out var result) ? result : Array.Empty<SetCookieHeaderValue>();
 
     internal static T? Get<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] T>(this IHeaderDictionary headers, string name)
     {
         ArgumentNullException.ThrowIfNull(headers);
+        var headerValue = headers[name];
 
-        var value = headers[name];
-
-        if (StringValues.IsNullOrEmpty(value))
+        if (StringValues.IsNullOrEmpty(headerValue))
         {
             return default(T);
         }
+        var value = headerValue.ToString();
 
-        if (KnownParsers.TryGetValue(typeof(T), out var temp))
+        // Comparison of typeof(T) is optimized by JIT.
+        if (typeof(T) == typeof(CacheControlHeaderValue))
+        {
+            return (T?)(object?)ParseCacheControlHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(ContentDispositionHeaderValue))
+        {
+            return (T?)(object?)ParseCacheContentDispositionHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(ContentRangeHeaderValue))
         {
-            var func = (Func<string, T>)temp;
-            return func(value.ToString());
+            return (T?)(object?)ParseCacheContentRangeHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(MediaTypeHeaderValue))
+        {
+            return (T?)(object?)ParseCacheMediaTypeHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(RangeConditionHeaderValue))
+        {
+            return (T?)(object?)ParseCacheRangeConditionHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(RangeHeaderValue))
+        {
+            return (T?)(object?)ParseCacheRangeHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(EntityTagHeaderValue))
+        {
+            return (T?)(object?)ParseCacheEntityTagHeaderValue(value);
+        }
+        else if (typeof(T) == typeof(DateTimeOffset?))
+        {
+            return (T?)(object?)ParseCacheDateTimeOffset(value);
+        }
+        else if (typeof(T) == typeof(long?))
+        {
+            return (T?)(object?)ParseCacheInt64(value);
```

Comment:
This casts a long to object. Does it box, or is this just ceremony to keep the compiler happy?

Reply:
It was to keep compiler happy. Not sure önt eh boxing, I do not recall seeing allocations. (cannot confirm as on the way to an airport)